File: javascript/026/main.js
/*
 * ProjectEuler 026
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v4.2.1
 * `node main.js`
 */

"use strict";

function main() {
    var max = 0;
    var dmax = 0;
    for (var d = 2; d < 1000; d++) {
        var r = rec(d);
        if (r > max) {
            max  = r;
            dmax = d;
        }
    }
    console.log(dmax);
}

function rec(d) {
    var o = Math.pow(10, Math.floor(Math.log(d) / Math.log(10)) + 1);
    var n = o;
    var memo = new Array(d + 1);
    while (true) {
        var j = Math.floor(n / d);
        if (memo[j] !== undefined) {
            return memo[j];
        }
        memo[j] = 0;
        n = (n % d) * o;
        if (n === 0) {
            return 0;
        }
        for (var i = 0; i <= d; i++) {
            if (memo[i] !== undefined) {
                memo[i]++;
            }
        }
    }
}


main();

File: ocaml/026/main.ml
(*
    ProjectEuler 026
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main main.ml`
*)

let recurring_cycle n =
    let memo = Array.make n (-1) in
    let rec itr a i =
        if a = 0 then 0
        else
            let j = Array.get memo a in
            if j >= 0 then i - j
            else begin
                Array.set memo a i;
                itr ((a * 10) mod n) (i + 1)
            end
    in
    let p = ceil (log10 (float_of_int n)) -. 1.0 in
    let e = int_of_float (10.0 ** p) in
    itr e 0

let main _ =
    let rec itr max_c max_d d =
        if d >= 1000 then max_d
        else
            let c = recurring_cycle d in
            if c > max_c then itr c d (d + 1)
            else itr max_c max_d (d + 1)
    in
    let d = itr 0 0 2 in
    print_int d;
    print_newline ()

let _ = main ()
